### DISCUSSION TITLE
Interface Not Registered

### DISCUSSION THREAD

**Discussion by RealCaseofgames** [dotnet/Silk.NET#1590] [Q&A] [answered]
I am trying to use D3D11 in Silk.NET, and I am following the sample given in this GitHub repository. However, I reach this line of code:
`SilkMarshal.ThrowHResult(factory.CreateSwapChainForHwnd(`
`    device,`
`    _renderWindow.WindowSurface.Native!.DXHandle!.Value,`
`    in scd,`
`    null,`
`    ref Unsafe.NullRef<IDXGIOutput>(),`
`    ref swapchain));`

When I reach it, the API throws a System.ArgumentException that reads "Interface not registered." I cannot figure out why. Is there something I am missing?

My Swap Chain is defined as:
`private ComPtr<IDXGISwapChain1> swapchain = default;`

I also define my Factory as:
`ComPtr<IDXGIFactory2> factory = dxgi.CreateDXGIFactory<IDXGIFactory2>();`

My Swap Chain Description is:
`SwapChainDesc1 scd = new();`
`scd.SampleDesc = new SampleDesc(1, 0);`
`scd.BufferUsage = DXGI.UsageRenderTargetOutput;`
`scd.BufferCount = 2;`
`scd.SwapEffect = SwapEffect.Discard;`
`scd.Flags = (uint)SwapChainFlag.AllowModeSwitch;`

I have also abstracted the window handle into its own class.

I am unsure if this is the right place to ask this question, but I don't really know where else to put it.

**Comment by Beyley** [comment]
Does the sample itself run?

**Reply by RealCaseofgames** [reply]
Yes. The sample itself runs.

**Comment by Perksey** [comment]
Do you have a more complete sample that we can just run/a more detailed stack trace? I've tried reproducing by tweaking a few parameters in the sample and I can't reproduce. You can also use the info queue to get more info:
```cs
// before you do anything with dxgi
using var infoQueue = dxgi.GetDebugInterface1<IDXGIInfoQueue>(0);
 
// check the info queue after failing to create the swapchain
var hr = factory.CreateSwapChainForHwnd(
    device,
    _renderWindow.WindowSurface.Native!.DXHandle!.Value,
    in scd,
    null,
    ref Unsafe.NullRef<IDXGIOutput>(),
    ref swapchain);
if (!HResult.IndicatesSuccess(hr))
{
    for (var i = 0ul; i < infoQueue.GetNumStoredMessages(); i++)
    {
        nuint msgByteLength = default; 
        SilkMarshal.ThrowHResult(infoQueue.GetMessageA(i, null, ref msgByteLength));
        var msgBytes = new byte[msgByteLength];
        ref var msg = ref Unsafe.As<byte, InfoQueueMessage>(ref msgBytes[0]);
        SilkMarshal.ThrowHResult(infoQueue.GetMessageA(i, ref msg, ref msgByteLength));
        Console.WriteLine(SilkMarshal.PtrToString((nint) msg.PDescription));
    }
}
```
Alternatively you can set up another thread to do this like we do in [D3D11's DebugHelpers](https://github.com/dotnet/Silk.NET/blob/main/src/Microsoft/Silk.NET.Direct3D11/DebugHelpers.cs).

Hope this helps.

**Reply by RealCaseofgames** [reply]
I tried to paste that into my code, but each method in the IDXGIInfoQueue requires a Producer parameter of type Guid.

**Comment by RealCaseofgames** [comment, marked as answer]
I just figured it out. I forgot to set the Format field in the SwapChainDesc1. The "Interface not registered" message was definitely misleading, and I think it should be changed to mirror what the struct _com_error in C++ does in interpreting an HRESULT.

**Reply by Perksey** [reply]
Excellent, glad you figured it out. There's nothing that Silk.NET can do to improve the experience here unfortunately, as this exception is raised from Marshal.ThrowExceptionForHR. It's worth noting I don't get your error on my machine when not changing the format field:
```
ID3D11Device::CreateTexture2D1: A Texture2D does not support the following Format (0, UNKNOWN).
ID3D11Device::CreateTexture2D1: Returning E_INVALIDARG, meaning invalid parameters were passed.
Unhandled exception. System.ArgumentException: Value does not fall within the expected range.
   at Silk.NET.Core.Native.SilkMarshal.ThrowHResult(Int32 hResult) in C:\Users\perks\Documents\_Silk.NET\Silk.NET\src\Core\Silk.NET.Core\Native\SilkMarshal.cs:line 803
   at Program.<>c__DisplayClass0_0.<<Main>$>g__OnLoad|0() in C:\Users\perks\Documents\_Silk.NET\Silk.NET\examples\CSharp\Direct3D11 Tutorials\Tutorial 1.2 - Hello quad\Program.cs:line 170
   at Silk.NET.Windowing.Internals.ViewImplementationBase.Initialize() in C:\Users\perks\Documents\_Silk.NET\Silk.NET\src\Windowing\Silk.NET.Windowing.Common\Internals\ViewImplementationBase.cs:line 108
   at Silk.NET.Windowing.WindowExtensions.Run(IView view) in C:\Users\perks\Documents\_Silk.NET\Silk.NET\src\Windowing\Silk.NET.Windowing.Common\WindowExtensions.cs:line 45
   at Program.<Main>$(String[] args) in C:\Users\perks\Documents\_Silk.NET\Silk.NET\examples\CSharp\Direct3D11 Tutorials\Tutorial 1.2 - Hello quad\Program.cs:line 94
```
